### PR TITLE
fix: Timestamp/range being re-opened on window resize.

### PR DESCRIPTION
**Pull request by peter-affenzeller**
### <strong>Pull Request</strong>

<hr>
Hi, thank you for contributing to Barista with this pull request (PR).

To ensure a fast process and merging of your PR please make sure it fulfills the
coding standards and contribution guidelines.

- A feature proposal has been provided, discussed and approved first.
- There is a meaningful description of the issue in GitHub (Screenshots are
  often helpful).
- If the PR introduces breaking-changes or deprecations it matches the following
  guidelines.
  - The commit message follows our commit guidelines.
  - Tests for the changes have been added (for bug fixes / features).
  - Docs have been added / updated (for bug fixes / features).

Please choose the type appropriate for the changes below: <br>

#### Type of PR

Bugfix (non-breaking change which fixes an issue)

#### Checklist

- [x] I have read the CONTRIBUTING doc and I follow the PR guidelines
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by tomheller** [comment]
@peter-affenzeller Our cirlceci trial has run out, which is why we cannot run builds on xlarge resources any more. Could you update your branch from the `dynatrace-oss/barista:master` to get the fix for this problem?

File: apps/components-e2e/src/components/chart/selection-area/selection-area.e2e.ts [hunk 1]
```diff
@@ -254,3 +254,49 @@ test('should create a range out of the timestamp with shift + left arrow key', a
...
+  await testController.expect(await rangeSelection.exists).notOk();
+  await testController.expect(await overlay.exists).notOk();
+
+  await testController.resizeWindow(1000, 600);
+
+  await testController.expect(await rangeSelection.exists).notOk();
+  await testController.expect(await overlay.exists).notOk();
+
+  await testController.resizeWindow(1200, 800);
+});
```
Comment (lukasholzer) [inline apps/components-e2e/src/components/chart/selection-area/selection-area.e2e.ts]: Please add a test that checks when a selection is open and you perform a resize of the window, that the selection is kept and not being closed.

Furthermore, it should select the same data (Check what is displayed in the overlay) 

Please add those tests for the range and the timestamp

File: components/chart/src/selection-area/selection-area.ts [hunk 2]
```diff
@@ -794,26 +800,43 @@ export class DtChartSelectionArea implements AfterContentInit, OnDestroy {
...
+        .subscribe(ref => {
+          if (this._chart._range && this._chart._range._overlayTemplate) {
+            this._updateOrCreateOverlay(this._chart._range, ref, undefined);
+          }
+        });
+
+      // update the overlay position if an overlay exists
+      // and the viewport size changes
+      combineLatest([
+        getElementRef<HTMLDivElement>(this._chart._range._rangeElementRef),
```
Comment (lukasholzer) [inline components/chart/src/selection-area/selection-area.ts]: this does not work out `getElementRef` returns not an observable it is a pipeable custom operator function.

**Comment by thomaspink** [comment]
As @lukasholzer said, please provide either a description in the PR or an issue.

File: components/chart/src/selection-area/selection-area.ts [hunk 3]
```diff
@@ -825,12 +848,30 @@ export class DtChartSelectionArea implements AfterContentInit, OnDestroy {
...
     // handling of the overlay for the timestamp
     if (this._chart._timestamp) {
+      getElementRefStream<HTMLDivElement>(
+        this._chart._timestamp._stateChanges,
+        this._destroy$,
+        this._chart._timestamp._timestampElementRef,
+        this._zone,
+      )
+        .pipe(
+          throttleTime(0, animationFrameScheduler),
```
Comment (lukasholzer) [inline components/chart/src/selection-area/selection-area.ts]: Please remove throttling here is no need for it. This stream does not fire that much

File: components/chart/src/selection-area/selection-area.ts [hunk 4]
```diff
@@ -825,12 +848,30 @@ export class DtChartSelectionArea implements AfterContentInit, OnDestroy {
...
+          }
+        });
+
+      // update the overlay position if an overlay exists
+      // and the viewport size changes
       combineLatest([
-        getElementRefStream<HTMLDivElement>(
-          this._chart._timestamp._stateChanges,
-          this._destroy$,
+        getElementRef<HTMLDivElement>(
```
Comment (lukasholzer) [inline components/chart/src/selection-area/selection-area.ts]: getElementRef cannot work here this is only a pipeable operator

**Comment by lukasholzer** [comment]
@peter-affenzeller please rebase with current master and resolve conflicts for further reviewing

File: components/chart/src/selection-area/selection-area.ts [hunk 5]
```diff
@@ -456,6 +457,11 @@ export class DtChartSelectionArea implements AfterContentInit, OnDestroy {
 
   /** If there is an overlay open it will dispose it and destroy it */
   private _closeOverlay(): void {
+    // remove class showing the arrows of the range handles
+    if (this._chart._range) {
+      this._chart._range._reflectRangeReleased(false);
```
Comment (lukasholzer) [inline components/chart/src/selection-area/selection-area.ts]: why is this needed?

File: apps/components-e2e/src/components/chart/selection-area/selection-area.e2e.ts [hunk 6]
```diff
@@ -244,3 +244,79 @@ test('should create a range out of the timestamp with shift + left arrow key', a
...
+    .resizeWindow(1000, 600)
+    .expect(timestampSelection.exists)
+    .ok()
+    .expect(overlay.exists)
+    .ok()
+    .expect(overlayText.textContent)
+    .match(/Jul 31, \d{2}:19/g);
+});
+
+test('should keep existing range open in case of a window resize', async (testController: TestController) => {
```
Comment (lukasholzer) [inline apps/components-e2e/src/components/chart/selection-area/selection-area.e2e.ts]: This test is flaky – your e2e test fails.

File: apps/components-e2e/src/components/chart/selection-area/selection-area.e2e.ts [hunk 7]
```diff
@@ -244,3 +244,79 @@ test('should create a range out of the timestamp with shift + left arrow key', a
...
+  await createRange(550, { x: 310, y: 100 }, testController)
+    .expect(rangeSelection.exists)
+    .ok()
+    .expect(overlayText.textContent)
+    .match(/Jul 31 \d{2}:17 — \d{2}:23/g)
+    .resizeWindow(1000, 600)
```
Comment (lukasholzer) [inline apps/components-e2e/src/components/chart/selection-area/selection-area.e2e.ts]: please add a wait after the resize

```suggestion
    .resizeWindow(1000, 600)
    .wait(500)
```

File: apps/components-e2e/src/components/chart/selection-area/selection-area.e2e.ts [hunk 8]
```diff
@@ -244,3 +244,79 @@ test('should create a range out of the timestamp with shift + left arrow key', a
...
+    chartClickTargets[1],
+    testController,
+  )
+    .expect(timestampSelection.exists)
+    .ok()
+    .expect(overlay.exists)
+    .ok()
+    .expect(overlayText.textContent)
+    .match(/Jul 31, \d{2}:19/g)
+    .resizeWindow(1000, 600)
```
Comment (lukasholzer) [inline apps/components-e2e/src/components/chart/selection-area/selection-area.e2e.ts]: please add wait after resizing

```suggestion
    .resizeWindow(1000, 600)
    .wait(500)
```

File: components/chart/src/selection-area/selection-area.ts [hunk 9]
```diff
@@ -803,17 +809,32 @@ export class DtChartSelectionArea implements AfterContentInit, OnDestroy {
 
     // handling of the overlay for the range
     if (this._chart._range) {
+      merge(
+        getElementRefStream<HTMLDivElement>(
```
Comment (lukasholzer) [inline components/chart/src/selection-area/selection-area.ts]: please create this stream and assign to a const you create the same stream some lines below

File: apps/components-e2e/src/components/chart/selection-area/selection-area.e2e.ts [hunk 10]
```diff
@@ -244,3 +244,79 @@ test('should create a range out of the timestamp with shift + left arrow key', a
...
+    .expect(rangeSelection.exists)
+    .ok()
+    .expect(overlay.exists)
+    .ok()
+    .click(closeButton)
+    .expect(rangeSelection.exists)
+    .notOk()
+    .expect(overlay.exists)
+    .notOk()
+    .resizeWindow(1000, 600)
```
Comment (lukasholzer) [inline apps/components-e2e/src/components/chart/selection-area/selection-area.e2e.ts]: please add a wait after the resize

File: apps/components-e2e/src/components/chart/selection-area/selection-area.e2e.ts [hunk 11]
```diff
@@ -244,3 +244,79 @@ test('should create a range out of the timestamp with shift + left arrow key', a
...
+    .notOk();
+});
+
+test('should not re-open a range in case of a window resize', async (testController: TestController) => {
+  await createRange(550, { x: 310, y: 100 }, testController)
+    .expect(rangeSelection.exists)
+    .ok()
+    .expect(overlay.exists)
+    .ok()
+    .click(closeButton)
```
Comment (lukasholzer) [inline apps/components-e2e/src/components/chart/selection-area/selection-area.e2e.ts]: please reduce speed with clicking

File: apps/components-e2e/src/components/chart/selection-area/selection-area.e2e.ts [hunk 12]
```diff
@@ -244,3 +244,79 @@ test('should create a range out of the timestamp with shift + left arrow key', a
...
+    .expect(timestampSelection.exists)
+    .ok()
+    .expect(overlay.exists)
+    .ok()
+    .click(closeButton)
+    .expect(timestampSelection.exists)
+    .notOk()
+    .expect(overlay.exists)
+    .notOk()
+    .resizeWindow(1000, 600)
```
Comment (lukasholzer) [inline apps/components-e2e/src/components/chart/selection-area/selection-area.e2e.ts]: please add a wait after the resize

File: apps/components-e2e/src/components/chart/selection-area/selection-area.e2e.ts [hunk 13]
```diff
@@ -244,3 +244,79 @@ test('should create a range out of the timestamp with shift + left arrow key', a
...
+  await createTimestamp(
+    { x: 450, y: 100 },
+    chartClickTargets[1],
+    testController,
+  )
+    .expect(timestampSelection.exists)
+    .ok()
+    .expect(overlay.exists)
+    .ok()
+    .click(closeButton)
```
Comment (lukasholzer) [inline apps/components-e2e/src/components/chart/selection-area/selection-area.e2e.ts]: please reduce speed

File: components/chart/src/selection-area/selection-area.ts (hunk 5 above)
Comment (tomheller) [reply components/chart/src/selection-area/selection-area.ts]: @peter-affenzeller I tried to finish your PR as requested, but when I removed this part of the code, the test failed. Did you figure out, why this is needed?

**Comment by tomheller** [comment]
@lukasholzer The flaky end2end test should be solved, when #610 is merged and rebased, as this deals with the update of the overlay, when change detection runs.

File: components/chart/src/selection-area/selection-area.ts (hunk 5 above)
Comment (peter-affenzeller) [reply components/chart/src/selection-area/selection-area.ts]: @lukasholzer @tomheller this is intended to remove the class that makes the little arrows of the drag handles fade in so that they fade in every time the range is opened (also with keyboard input. currently, when you focus the chart and convert a timestamp to a range they don't show), in combination with the reflectRangeReleased(true) in line 334. but I just checked again, this also seems to not work properly, probably better to do this in another PR.

**Comment by tomheller** [comment]
Pull request commit `732cb40b914854890575fceb8e878a3d32c24f66` has been cherry picked to 5.x and 5.2.x
